### ISSUE TITLE
conda build conda.recipe --python=2.7 builds against Python 3.5 on datashape's build

### ISSUE THREAD

**Issue by cpcloud** [conda/conda-build#704]
The package spec is currently `python 2.7*|3.4*|3.5*` and when I change it to just `python`, the above line works and tests against the correct python version. 

Here's the [`meta.yaml`](https://github.com/blaze/datashape/blob/master/conda.recipe/meta.yaml).

**Comment by cpcloud** [comment]
This spec also breaks `python >=2.7,<3.3|>=3.4`

**Comment by cpcloud** [comment]
It seems to default to 3.5 whenever it's allowed

**Comment by cpcloud** [comment]
I can reproduce this on the `pyzmq` recipe

**Comment by ccordoba12** [comment]
Are you using this command?

    conda build conda.recipe --python=2.7

Because I think it should be

    conda build --python=2.7 conda.recipe

I didn't try it with the `datashape` recipe, but I've used it for things like `boost` and it works fine for me :-)

**Comment by cpcloud** [comment]
The order of arguments doesn't matter, the argument parser ignores the order of named arguments. This true of any command line program written with `argparse`

**Comment by cpcloud** [comment]
@ccordoba12 You'll only be able to reproduce this bug if you have a `python` specification that *explicitly* allows Python 3.5. For example, the follow package spec builds the version that I pass in to `--python`:

```yaml
build:
  - python
```

And the next one **fails** to build against Python 2.7, even thought that's what I passed *and* it matches the spec:

```yaml
build:
  - python >=2.7,<3.3|>=3.4  # <- this always builds against Python 3.5 no matter what
```

**Comment by ccordoba12** [comment]
> The order of arguments doesn't matter

Ok, I wasn't aware of that. I thought that could be the problem.

> You'll only be able to reproduce this bug if you have a python specification that explicitly allows Python 3.5

Then this seems a bug in the solver, or in the way conda-build sets `--python`. I think the last one is the most probable problem. I'll investigate a little bit to see if I can find the cause.

**Comment by stuarteberg** [comment]
I recommend the following syntax in your `meta.yaml` file.

```yaml
build:
  - python {{PY_VER}}*,>=2.7,<3.3|>=3.4
```

Either of the following commands will work:

- `conda build --python=2.7 datashape`
- `CONDA_PY=27 conda build datashape`

Thus, the presence of `{{PY_VER}}*` will pin python to whatever you asked for on the command line, but not all values of `--python` will be accepted, depending on the remainder of the spec (`>=2.7,<3.3|>=3.4`).

**Comment by whipit** [comment]
New to conda -- but is "conda-build" and "conda build" the same thing?

**Comment by msarahan** [comment]
@whipit yes, conda-build is the actual executable, and ``conda build`` is a command that ends up resolving to conda-build.

Please don't ask questions on the tail end of unrelated issues, please.

**Comment by djhoese** [comment]
FYI I have the same issue where my `meta.yaml` uses `python >=3.6` and now that 3.7 is available it was building for 3.7 instead of 3.6. Specifying `--python "3.6"` didn't work until I followed @stuarteberg's example.

**Comment by msarahan** [comment]
If you specify python constraints, conda-build's ability to pass in its own constraints is overridden.  @stuarteberg's example is clever, because it allows conda-build to pass in the necessary PY_VER, which obviates the need for any sort of additional constraint.

It is best to use @stuarteberg's example, or not specify a python constraint at all.
